### ISSUE TITLE
Desychronization of retroknowledge in module types

### ISSUE THREAD

**Issue by herbelin** [coq/coq#14822]
#### Description of the problem

```
Module Type M. Primitive float := #float64_type. End M.
Declare Module Import N : M.
Check float. (* ok *)
Primitive sqrt := #float64_sqrt. (* anomaly *)
```

Comparatively, both
```
Primitive float := #float64_type.
Check float. (* ok *)
Primitive sqrt := #float64_sqrt. (* ok *)
```
and
```
Module Import M. Primitive float := #float64_type. End M.
Check float. (* ok *)
Primitive sqrt := #float64_sqrt. (* ok *)
```
work.

Note that the anomaly is #14814 but even with #14814 fixed there would be an error showing a desynchronization.

I believe that module types should fully support declarating retroknowledge?? Cc @ppedrot.

#### Coq Version

Since introduction of floats in 8.11.

**Comment by ppedrot** [comment]
Adding retroknowledge to higher-order structures looks like playing with fire to me. This should only be allowed inside toplevel modules instead.

**Comment by herbelin** [comment]
What are the risks other than an error telling that retroknowledge is already registered for such or such primitive?

**Comment by ppedrot** [comment]
This is an imperative mechanism, so weird soundness bugs / crashes in the VM.

**Comment by herbelin** [comment]
You mean in interactive mode with backtracking?

**Comment by herbelin** [comment]
Another example showing the converse desynchronization:
```
Module Type T. End T.
Module Import M : T. Primitive float := #float64_type. End M.
Fail Check float.
Primitive sqrt := #float64_sqrt. (* success *)
Print sqrt.
(* Anomaly "Constant Top.M.float does not appear in the environment." *)
```

**Comment by herbelin** [comment]
Another example showing that the current approach does not prevent using `Primitive` in functors:
```
Module Type T. End T.
Module Type U (X:T). End U. 
Module F (X:U).
  Module M : T. Primitive float := #float64_type. End M.
  Module N := X M.
End F.
Module P (X:T). End P.
Module Q := F P. (* does side effect on declaring #float64_type *)
Primitive sqrt := #float64_sqrt. (* then this is ok *)
Print sqrt. (* also, float is still not there *)
(* Anomaly "Constant Top.F.M.float does not appear in the environment." *)
```

**Comment by herbelin** [comment]
In comparison, OCaml accepts to bind the same `external` to several names. Indeed, the primitive is there in the VM anyway. What we need is a name to refer to it. We could give it several names, using the nametab to decide which name to use when we need to print it.

**Comment by ppedrot** [comment]
Please stop right now, I am getting shingles. Retroknowledge is just a global registering mechanism, and the Coq world is not comparable to the OCaml world. We're not going to pile fragile hacks in the kernel to handle conversion of potentially different externals when we only have a handful of them that should be instantiated only a single time anyways. Just forbid those primitive declarations by default, except in very specific cases.

**Comment by herbelin** [comment]
I doubt a colorful language might help, moreover when piling wrong claims :smile:. Please rather propose a concrete solution.

**Comment by ppedrot** [comment]
> This should only be allowed inside toplevel modules instead.

**Comment by herbelin** [comment]
You want to disallow also `Module M := Floats`?

**Comment by ppedrot** [comment]
Yes if `Floats` would contain retroknowledge registrations, but it's not the case. We already have minimal modules `PrimFoo` declaring the necessary retroknowledge registrations. I don't think any sane person would build module indirections with those.

**Comment by herbelin** [comment]
So, your proposal is to move retroknowledge out of `module_body` and to have it only in `compiled_library` + forbidding including or aliasing a library containing retroknowledge to a module?

**Comment by ppedrot** [comment]
Indeed. Instead of storing the retroknowledge in the module body, just keep a boolean indicating that it was tainted and cannot be aliased, and do everything else at the level of the library, which is the global scope in which registrations live.

**Comment by herbelin** [comment]
Personally, I still find more principled the view that there is a constructor in `constr` for each primitive object. This is actually morally the case except that this constructor is called `Declarations.Primitive` and is accessed only indirectly via a constant name that is declared dynamically. The current approach pretends that side-effects are needed where I don't actually see the need for.

**Comment by ppedrot** [comment]
> The current approach pretends that side-effects are needed

It does not really use side-effects, rather it's delayed link (a.k.a. lazy computation). In particular you shouldn't be able to overwrite the value of registered constants if they've been registered once somewhere in the chain of dependencies.

You can probably design a system where you can register several different instances of the same underlying primitive function, but it's going to be hard to check that it's correct. You'll notably have to keep unique kernel names in the primitive itself to separate the various instances, so that conversion is sound. So, quite a heavy system for dubious actual use. If you consider atop of that the usual horrors of module aliasing and the like, this leaves a lot of room for interesting behaviours (read soundness issues).

**Comment by herbelin** [comment]
@ppedrot, does your proposal roughly amount to:
- to remove `mod_retroknowdege` from `module_mody`
- to remove `Primitive` from `constant_def` (but keep it in `Environ.Globals.constant`)
- to have a proper `Safe_typing.add_primitive` binding to a simplification of `Term_typing.infer_primitive` with global registration to a library but not to a module, checking that no module other than a library one is open

That would then simplify the module management.
